rfd/0163-vnet.md

```diff
@@ -0,0 +1,326 @@
+    1. If no match is found and the zone has no `upstream_nameservers` configured,
+       no answer will be sent.
+    1. The full DNS query will be forwarded to one of the `upstream_nameservers`,
+       selected randomly from the list, and any upstream response will be
+       returned to the client verbatim.
+       We may implement caching of upstream responses, but hopefully the
+       downstream resolver will do its own caching.
+1. If the name did not match any `<proxy-address>.internal` or custom DNS zone,
+   no response will be sent.
+
+TODO: sequence diagram
+
+### IP range
+
+By default IPs will be assigned from the carrier-grade NAT range 100.64.0.0/10
+Since VNet IPs will only be used locally on end user machines this range is
+quite unlikely to conflict with any valid public or private IP addresses, unless
+another VPN client is being used that happens to use the same range.
+For example, tailscale also uses this range.
+This range can be overriden with a cluster-wide configuration.
+
+```yaml
+version: v1
+kind: vnet
+metadata:
+  name: vnet
+spec:
+  cidr_range: 100.64.0.0/10
+```
+
+### Host configuration
+
+Vnet will need to create a TUN interface, assign an IP to it, and modify the
+routing table.
+
+To create the TUN interface we will import `golang.zx2c4.com/wireguard/tun`
+which provides cross-platform methods for creating TUN interfaces and reading
+from + writing them.
+It supports MacOS, Windows, and Linux.
+
+On MacOS, to assign an IP address to the interface, we will use `os/exec` to run
+the `ifconfig` command:
+
+```shell
+ifconfig utun4 100.64.0.1 100.64.0.1 up
+```
+
+On MacOS, to assign an IP route to use the TUN interface, we will use `os/exec`
+to run the `route` command:
+
+```shell
+route add -net 100.64.0.0/10 -interface utun4
+```
+
+All system configuration must run as a privileged user/root.
+In the first version we will use an Applescript wrapper to call a child process
+as root.
+The child admin process will still be the `tsh` binary, just called with
+arguments to do the host setup.
+See the UX section for more details on the admin process.
+
+The first version of VNet will only support MacOS so this is written with that
+in mind, but we can implement this in a similar way for Windows and Linux.
+
+### Security
+
+### UX
+
+VNet is available in both tsh and Teleport Connect. This section focuses mostly on VNet integration
+in Teleport Connect, with a short subsection on VNet in tsh at the end.
+
+Attempting to connect to a TCP app opens the VNet tab instead. When the VNet app is opened this way,
+it gives the user an option to either start VNet or open a standard local proxy instead.
+
+Web apps continue to launch through the proxy service. The three dot menu next to web apps gains a
+new action called "Open through VNet". If VNet is running, it opens the app VNet address in the
+browser. If VNet is not running, it opens the VNet tab saying "You can open dumper and other apps
```

Comment:
> I don't _think_ we really ever want to open a VNet address in the browser, (…) We can show the VNet address for HTTP apps somewhere because it still works, especially for REST APIs, but the big usecase will be TCP apps I think.

Right, totally. The "Open through VNet" action would be under the three dots menu, just like "Set up connection".

<img width="468" alt="set-up-connection" src="https://github.com/gravitational/teleport/assets/27113/58aa403d-7692-4b4d-9fe4-1294f4866e6f">

I'll give it another go tomorrow. The more I think about it, the more I come to the conclusion that we perhaps shouldn't even call it VNet within the app. Like we could just call it "local app proxy" or whatever and the only thing the user would have to know is that if the local app proxy is active, they can access the app through a special address with no extra authentication on their part.

IIRC, from the perspective of App Access there's no way to distinguish between HTTP apps and HTTP APIs, so the UI will have to work with both.

Reply:
Many customers may prefer VNet address if it matches their DNS, in this case the experience for most users is that they won't notice that they are using the tunnel.

Reply:
I've been thinking about this more. I agree using the app's public address in VNet would be a much better UX, and I believe we can make it work.

For a cluster `teleport.example.com` VNet will configure itself as the local DNS resolver for `*.teleport.example.com` and any custom DNS zones.

For a TCP app `api.teleport.example.com` or `api.custompublicaddr.example.com` it's fairly simple, we will always proxy TCP all the way to the app over the authenticated tunnel.

For an HTTP app `grafana.teleport.example.com` or `grafana.custompublicaddr.example.com` I think we should make sure it still works when opened from the web UI the same way it does today, with TLS terminated at the proxy. We we can make it *also* work when connecting over plain HTTP.

VNet can do this by peeking at the first few bytes of the incoming TCP connection. If it's a TLS client hello, forward TCP straight to the proxy. Assuming it was opened in the browser client auth will happen via the session cookie the same way it does without VNet. If it's not TLS, we'll assume the client is sending plain HTTP and send it over an mTLS authenticated tunnel to the app.

I'll write this up in the RFD tomorrow

Reply:
Two things that come to my mind:

1. Will this require root privileges when stopping VNet to remove VNet as the DNS resolver? If someone quits Connect while VNet was running, I imagine we'd like `*.teleport.example.com` to not go through the DNS resolver which is no longer running.
2. Because our Web UI sends `Strict-Transport-Security: max-age=31536000; includeSubDomains`, opening the HTTP version of web apps in the browser will not be easy, depending on the browser. The user will need to clear HSTS from the domain. Granted, I understand this is something we don't want to encourage – web apps should be used through the proxy service, HTTP APIs should be used through VNet.
3. If an HTTP API is protected with TLS and whatever app I'm working on is hardcoded to connect to it over HTTPS, doesn't it mean that it'll be impossible to use VNet here without changing the protocol to HTTP?
   * I feel like this is a scenario where we could use some input from the design partner and other companies. The best course of action here depends on how their codebases look like right now.

Reply:
You're right on (1), but I don't think the problem is limited to this scenario. If a company has a custom DNS zone `example.com` they probably don't want everything under that to break when Connect exits either. I'm really starting to think we should leave the root process running in the background. We can communicate any required changes with it, and it can clean up the configuration before it exits.

For (2) I think it's okay, just don't try to open the HTTP version in the browser. curl and scripts should still work fine

For (3), you already can't connect to HTTP Apps with regular TLS, you have to use the session cookies or mTLS with app certs, so there's no breaking change here. HTTPS behaviour will be the same as it is today, HTTP will work, and If you want to do TLS terminated at the app you can use a TCP app

Reply:
Messing with this today and there's a challenge I didn't think of at first. I was planning to configure vnet as the DNS nameserver for only `*.teleport.example.com` by creating a file `/etc/resolver/teleport.example.com`, but this also configures it as the nameserver for the proxy address `teleport.example.com` and that becomes a problem when vnet itself tries to find the proxy address.

I think we'll need to discover the default/non-vnet nameservers on the system and forward requests for the proxy address to those. If we do this then we probably don't need to make the user configure upstream nameservers for custom dns zones either

Reply:
Well I was thinking the root process lifetime would still be tied to the Connect lifetime, it just wouldn't immediately exit after the initial setup. We could keep the unix socket we're already using open and communicate over that.

But now that I think of it, there is also a potential world where a VNet daemon starts as root at boot time and then it actually opens Connect with a deep link when it needs a login. This would be much more similar to a VPN, especially if we put it in the dock. Could be a phase 2 thing.

For the other point, if they want TLS to their app to work like a VPN, from a script without mTLS or requiring a web UI login, then their app should be terminating TLS (like it presumably was before they adopted teleport) and they should register it as a TCP app and connect over vnet

Reply:
As for leaving a root process running and communicating with it over a socket, we're running into a similar problem as described in the `SMAppService` section – how do we make sure no one else can talk to the privileged process?

With XPC, [the deamon can verify if the connecting process has been signed by the same Apple dev cert](https://theevilbit.github.io/posts/secure_coding_xpc_part1/#fix-such-and-similar-issues). I found [a Go package](https://github.com/xorrior/xpcutil) that can seemingly send messages to an XPC service, but it does not implement starting an XPC service. AFAIK an XPC service needs to be started from a launch daemon anyway, which would lead us down the `SMAppService` path.

Without XPC, the only idea that comes to my mind is how Connect handles lack of Unix socket support in Node.js on Windows. Each process generates a self-signed cert, saves the public part to a predetermined location accessible only by the current user (the dir is passed over argv from the parent process to the child process), and then they both use the certs for gRPC with mTLS.

If there's a way to avoid doing the cleanup as root, it'd greatly reduce the effort needed to release v1.

Reply:
I think we're okay because the privileged process is actually dialing a socket created by the non-privileged process

And I thought we might, but actually I don't think we need to do any additional communication. The root process can just keep running in the background and do 2 things:
1. watch the TSH_HOME for any clusters we logged in to and update the DNS configuration accordingly
2. watch the original parent process to see when it dies, then clean up and exit

Reply:
This thread is getting very long, I think the original questions are resolved, we can continue the root process lifetime discussion here https://github.com/gravitational/teleport/pull/37834#discussion_r1509872421